### PR TITLE
[Azure Blob Storage] Optimize memory consumption

### PR DESCRIPTION
**Pull request by moxarth-elastic**
## Closes https://github.com/elastic/connectors/issues/2465

Improve memory consumption in ABS connector. Added perf8 reports for the comparision:

Before: https://drive.google.com/file/d/12fcQWT8NKUr9CMgdEZsqVscXLEiFNJw-/view?usp=drive_link
After: https://drive.google.com/file/d/1IOKhm2s8hk0abjGarFy75d64AmKeANYE/view?usp=drive_link

## Fix:

The previous version of the connector instantiated a BlobClient for each blob it downloaded. Where, the new approach utilizes a single ContainerClient to download data for all blobs within that container. 

For instance, if a container contains 100 blobs, the old code was creating 100 separate BlobClient instances to download those blobs, but now the new code leverages just one ContainerClient for all blobs in the container. This optimization was inspired by a suggestion from[ this GitHub issue](https://github.com/Azure/azure-sdk-for-python/issues/27023#issuecomment-1314296101).

Additionally, the older code used an asynciterator with the data.chunks() method. The drawback of asynciterator is that it does not release memory until it has completed all iterations.

#### Pre-Review Checklist
- [x] this PR does NOT contain credentials of any kind, such as API keys or username/passwords (double check `config.yml.example`)
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

#### Changes Requiring Extra Attention

- Updated the version for `azure-storage-blob` in the requirements.

### CODE REVIEW

**Review by artem-shelkovnikov — COMMENTED** [review]
Can you add description to the PR why this PR improves RAM usage? What were your findings?

**Comment by moxarth-elastic** [comment]
> Can you add description to the PR why this PR improves RAM usage? What were your findings?

Added a fix in the description.

**Comment by moxarth-elastic** [comment]
Buildkite test this

File: connectors/sources/azure_blob_storage.py [hunk 1]
```diff
@@ -190,19 +191,38 @@ async def get_content(self, blob, timestamp=None, doit=None):
...
+                retry_total=self.retry_count,
+            )
+            return self.container_clients[container_name]
+        else:
+            return self.container_clients[container_name]
+
+    async def blob_download_func(self, blob_name, container_name, file_size):
+        container_client = self._get_container_client(container_name=container_name)
+        offset = 0
+        length = 100000
```
Comment (artem-shelkovnikov) [inline connectors/sources/azure_blob_storage.py:213]: What is this number? Why 100000? Can we put it into a meaningful constant or even a setting?

File: connectors/sources/azure_blob_storage.py [hunk 2]
```diff
@@ -190,19 +191,38 @@ async def get_content(self, blob, timestamp=None, doit=None):
...
-        async with BlobClient.from_connection_string(
-            conn_str=self.connection_string,
-            container_name=container_name,
-            blob_name=blob_name,
-            retry_total=self.retry_count,
-        ) as blob_client:
-            data = await blob_client.download_blob()
-            async for content in data.chunks():
-                yield content
+    def _get_container_client(self, container_name):
```
Comment (artem-shelkovnikov) [inline connectors/sources/azure_blob_storage.py:197]: Will you need to close these clients manually?

**Comment by moxarth-elastic** [comment]
buildkite test this

File: connectors/sources/azure_blob_storage.py (hunk 2 above)
Comment (moxarth-elastic) [reply connectors/sources/azure_blob_storage.py:205]: Yes, I've added a `close` method for the same.

**Comment by moxarth-elastic** [comment]
buildkite test this

File: connectors/sources/azure_blob_storage.py (hunk 1 above)
Comment (moxarth-elastic) [reply connectors/sources/azure_blob_storage.py]: It's used to fetch the file content in the chunks of 100 Kb. I've put it in a constant.

**Review by artem-shelkovnikov — APPROVED** [review]

**Comment by praveen-elastic** [comment]
Buildkite test this

**Comment by praveen-elastic** [comment]
Buildkite test this

**Pull request merged**